libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1074 @@
+                                                                             callback,
+                                                                             user_data);
+  guint i = 0;
+  g_autoptr(GPtrArray) word_proxies = g_ptr_array_new ();
+  g_autoptr(GPtrArray) quote_proxies = g_ptr_array_new ();
+  g_autoptr(GPtrArray) zipped_word_quote_proxies = NULL;
+
+  for (i = 0; i < ka_proxies->len; ++i)
+    {
+      EosDiscoveryFeedKnowledgeAppProxy *ka_proxy = g_ptr_array_index (ka_proxies, i);
+      GDBusProxy *dbus_proxy = eos_discovery_feed_knowledge_app_proxy_get_dbus_proxy (ka_proxy);
+      const gchar *interface_name = g_dbus_proxy_get_interface_name (dbus_proxy);
+
+      if (g_strcmp0 (interface_name, "com.endlessm.DiscoveryFeedContent") == 0)
+        append_stores_task_from_proxy (ka_proxy,
+                                       append_discovery_feed_content_from_proxy,
+                                       (GDestroyNotify) g_ptr_array_unref,
+                                       append_discovery_feed_content_from_proxy_data_new ("ArticleCardDescriptions",
+                                                                                          EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTICLE_CARD,
+                                                                                          EOS_DISCOVERY_FEED_CARD_LAYOUT_DIRECTION_IMAGE_FIRST,
+                                                                                          eos_discovery_feed_knowledge_app_card_store_new,
+                                                                                          EOS_DISCOVERY_FEED_THUMBNAIL_SIZE_ARTICLE),
+                                       cancellable,
+                                       individual_task_result_completed,
+                                       individual_task_result_closure_new (all_tasks_results_closure_allocate_new_result_slot (all_tasks_closure),
+                                                                           all_tasks_closure));
+      else if (g_strcmp0 (interface_name, "com.endlessm.DiscoveryFeedNews") == 0)
+        append_stores_task_from_proxy (ka_proxy,
+                                       append_discovery_feed_content_from_proxy,
+                                       (GDestroyNotify) g_ptr_array_unref,
+                                       append_discovery_feed_content_from_proxy_data_new ("GetRecentNews",
+                                                                                          EOS_DISCOVERY_FEED_CARD_STORE_TYPE_ARTICLE_CARD,
+                                                                                          EOS_DISCOVERY_FEED_CARD_LAYOUT_DIRECTION_IMAGE_LAST,
+                                                                                          (EosDiscoveryFeedKnowledgeAppCardStoreFactoryFunc) eos_discovery_feed_knowledge_app_news_card_store_new,
```

Comment:
Can we find a way to avoid the cast here?

Reply:
I could make `eos_discovery_feed_knowledge_app_news_card_store_new` return a `EosDiscoveryFeedKnowledgeAppCardStore` (its parent type), sort of like how `gtk_window_new` returns a `GtkWidget`. Not sure if we want to do that though, or if the cast is better.